text/0000-build-systems.md

```diff
@@ -0,0 +1,455 @@
+- Feature Name: N/A
+- Start Date: 2017-09-01
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This **experimental RFC** lays out a high-level plan for improving Cargo's
+ability to integrate with other build systems and environments. As an
+experimental RFC, it opens the door to landing [unstable features] in Cargo to
+try out ideas, but *not* to stabilizing those features, which will require
+follow-up RFCs. It proposes a variety of features which, in total, permit a wide
+spectrum of integration cases -- from customizing a single aspect of Cargo to
+letting an external build system run almost the entire show.
+
+[unstable features]: https://github.com/rust-lang/cargo/pull/4433/
+
+# Motivation
+[motivation]: #motivation
+
+One of the first hurdles for using Rust in production is integrating it into
+your organization's build system. The level of challenge depends on the level of
+integration required: it's relatively painless to invoke Cargo from a makefile
+and let it fully manage building Rust code, but gets harder as you want the
+external build system to exert finer-grained control over how Rust code is built.
+The goal of this RFC is to lay out a vision for making integration at *any* scale
+much easier than it is today.
+
+After extensive discussion with stakeholders, there appear to be two distinct
+kinds of use-cases (or "customers") involved here:
+
+- **Mixed build systems**, where building already involves a variety of
+  language- or project-specific build systems. For this use case, the desire is
+  to use Cargo as-is, except for some specific concerns. Those concerns take a
+  variety of shapes: customizing caching, having a local crate registry, custom
+  handling for native dependencies, and so on. Addressing these concerns well
+  means adding new points of extensibility or control to Cargo.
+
+- **Homogeneous build systems** like [Bazel], where there is a single prevailing
+  build system and methodology that works across languages and projects and is
+  expected to drive all aspects of the build. In such cases the goal of Cargo
+  integration is largely *interoperability*, including easy use of the crates.io
+  ecosystem and Rust-centric tooling, both of which expect Cargo-driven build
+  management.
+
+[Bazel]: https://bazel.build/
+
+The interoperability constraints are, in actuality, hard constraints around
+*any* kind of integration.
+
+In more detail, a build system integration *must*:
+
+- Make it easy for the outer build system to control the aspects of building
+  that are under its purview (e.g. artifact management, caching, network access).
```

Comment:
bazel is also highly interested in sandboxing, so it wants to control disk paths and not allow general file system access. This means, for bazel to call cargo, all the paths (tmp directories, working directories, any cache directories that cargo uses) should be configurable by command or config file that bazel can pass in.

Additionally, bazel does not love keeping track of stateful directories (in fact, it really can't very well). It wants files. So were possible, the ability to consume and emit zip files instead of directories would be nice.

Finally, bazel is into reproducibility and uses hashes of outputs to see if downstreams need to be recompiled. Outputs of cargo/rustc ideally can disable any timestamping into the binary or any non-reproducible aspect of artifact production.

Reply:
>  Outputs of cargo/rustc ideally can disable any timestamping into the binary or any non-reproducible aspect of artifact production.

https://github.com/rust-lang/rust/issues/34902